File: scripts/quiz/src/selection.ts
import { shuffle } from "./random.js";
import type {
  FilterCriterion,
  Question,
  SetBundle,
  SetConstraint,
  SetFilters,
  SetSelectionFixed,
  SetSelectionQuery,
  TaxonomyCatalog
} from "./types.js";

export const SELECTION_ALGORITHM_VERSION = 1;

export class SelectionError extends Error {}

export function* combinations<T>(items: readonly T[], count: number): Generator<T[]> {
  if (count <= 0) {
    yield [];
    return;
  }
  if (count > items.length) {
    return;
  }
  const n = items.length;
  const indices = Array.from({ length: count }, (_, i) => i);
  yield indices.map(i => items[i]);

  while (true) {
    let i = count - 1;
    while (i >= 0 && indices[i] === i + n - count) {
      i -= 1;
    }
    if (i < 0) {
      return;
    }
    indices[i] += 1;
    for (let j = i + 1; j < count; j += 1) {
      indices[j] = indices[j - 1] + 1;
    }
    yield indices.map(idx => items[idx]);
  }
}

export function getTopicDescendantsAndSelf(topicId: string, taxonomy?: TaxonomyCatalog): Set<string> {
  const result = new Set<string>([topicId]);
  if (!taxonomy || !taxonomy.topics) return result;

  let changed = true;
  while (changed) {
    changed = false;
    for (const [tid, topic] of Object.entries(taxonomy.topics)) {
      if (topic.parent && result.has(topic.parent) && !result.has(tid)) {
        result.add(tid);
        changed = true;
      }
    }
  }
  return result;
}

function matchesAnyOrAll(itemValues: readonly string[], criterion: FilterCriterion): boolean {
  if ("any" in criterion && Array.isArray(criterion.any)) {
    const target = new Set(criterion.any);
    return itemValues.some(v => target.has(v));
  }
  if ("all" in criterion && Array.isArray(criterion.all)) {
    const itemSet = new Set(itemValues);
    return criterion.all.every(v => itemSet.has(v));
  }
  return true;
}

export function matchesFilters(question: Question, filters?: SetFilters, taxonomy?: TaxonomyCatalog): boolean {
  if (!filters) return true;

  // 1. Topics
  if (filters.topics) {
    const qTopics = question.topicIds ?? (question as unknown as { topics?: string[] }).topics;
    if (!qTopics || !Array.isArray(qTopics) || qTopics.length === 0) return false;

    if ("any" in filters.topics && Array.isArray(filters.topics.any)) {
      const expanded = new Set<string>();
      for (const t of filters.topics.any) {
        for (const desc of getTopicDescendantsAndSelf(t, taxonomy)) {
          expanded.add(desc);
        }
      }
      if (!qTopics.some(t => expanded.has(t))) return false;
    } else if ("all" in filters.topics && Array.isArray(filters.topics.all)) {
      for (const t of filters.topics.all) {
        const expanded = getTopicDescendantsAndSelf(t, taxonomy);
        if (!qTopics.some(qt => expanded.has(qt))) return false;
      }
    }
  }

  // 2. Concepts
  if (filters.concepts) {
    const qConcepts = question.conceptIds ?? (question as unknown as { concepts?: string[] }).concepts;
    if (!qConcepts || !Array.isArray(qConcepts) || qConcepts.length === 0) return false;
    if (!matchesAnyOrAll(qConcepts, filters.concepts)) return false;
  }

  // 3. Objectives
  if (filters.objectives) {
    const qObjs = question.objectiveIds ?? (question as unknown as { objectives?: string[] }).objectives;
    if (!qObjs || !Array.isArray(qObjs) || qObjs.length === 0) return false;
    if (!matchesAnyOrAll(qObjs, filters.objectives)) return false;
  }

  // 4. Related pages
  if (filters.related_pages) {
    const qPages = question.relatedPages ?? (question as unknown as { related_pages?: string[] }).related_pages;
    if (!qPages || !Array.isArray(qPages) || qPages.length === 0) return false;
    if (!matchesAnyOrAll(qPages, filters.related_pages)) return false;
  }

  // 5. Types
  if (filters.types && Array.isArray(filters.types)) {
    if (!filters.types.includes(question.type)) return false;
  }

  // 6. Cognitive levels
  if (filters.cognitive_levels && Array.isArray(filters.cognitive_levels)) {
    const cog = question.cognitiveLevel ?? (question as unknown as { cognitive_level?: string }).cognitive_level;
    if (!cog || !filters.cognitive_levels.includes(cog)) return false;
  }

  // 7. Styles
  if (filters.styles && Array.isArray(filters.styles)) {
    if (!question.style || !filters.styles.includes(question.style)) return false;
  }

  // 8. Question IDs
  if (filters.question_ids && Array.isArray(filters.question_ids)) {
    if (!filters.question_ids.includes(question.id)) return false;
  }

  // 9. Difficulty
  if (filters.difficulty) {
    const diff = question.difficulty;
    if (diff === undefined || typeof diff !== "number") return false;
    if (filters.difficulty.min !== undefined && diff < filters.difficulty.min) return false;
    if (filters.difficulty.max !== undefined && diff > filters.difficulty.max) return false;
  }

  return true;
}

export function satisfiesConstraints(selected: readonly Question[], constraints?: readonly SetConstraint[]): boolean {
  if (!constraints || constraints.length === 0) return true;

  for (const c of constraints) {
    const values = c.values ?? [];
    let count = 0;
    for (const q of selected) {
      const val = (q as unknown as Record<string, unknown>)[c.field] ?? (c.field === "type" ? q.type : undefined);
      if (values.includes(val as string | number)) {
        count += 1;
      }
    }
    if (c.min !== undefined && count < c.min) return false;
    if (c.max !== undefined && count > c.max) return false;
  }

  return true;
}

export function withShuffledChoices(question: Question, setId: string, seed?: string): Question {
  if (seed === undefined) return question;
  const choiceOrder =
    question.choiceOrder ?? (question as unknown as { choice_order?: "shuffle" | "fixed" }).choice_order ?? "shuffle";
  if (choiceOrder !== "shuffle") return question;

  if ((question.type === "single_choice" || question.type === "multiple_choice") && question.choices) {
    return {
      ...question,
      choices: shuffle(question.choices, `${setId}:${question.id}:choices:${seed}`)
    };
  }
  return question;
}

export function canSatisfyConstraints(
  selected: readonly Question[],
  remainingCount: number,
  constraints?: readonly SetConstraint[]
): boolean {
  if (!constraints || constraints.length === 0) return true;
  for (const c of constraints) {
    const values = c.values ?? [];
    let count = 0;
    for (const q of selected) {
      const val = (q as unknown as Record<string, unknown>)[c.field] ?? (c.field === "type" ? q.type : undefined);
      if (values.includes(val as string | number)) {
        count += 1;
      }
    }
    if (c.max !== undefined && count > c.max) return false;
    if (c.min !== undefined && count + remainingCount < c.min) return false;
  }
  return true;
}

export type InfeasibleReasonCode =
  | "insufficient_pool"
  | "slot_insufficient_candidates"
  | "slot_conflict"
  | "constraint_violation";

export type SelectionOutcome =
  | { status: "ok"; questions: Question[] }
  | { status: "infeasible"; reasonCode: InfeasibleReasonCode; message: string }
  | { status: "exhausted"; message: string };

export function solveQuerySelectionDetailed(
  pool: readonly Question[],
  query: SetSelectionQuery,
  taxonomy?: TaxonomyCatalog,
  seed?: string,
  setId = ""
): SelectionOutcome {
  const count = query.count;
  const topFilters = query.filters;
  const slots = query.slots ?? [];
  const constraints = query.constraints ?? [];

  // Step 1: Filter pool by top-level filters and sort by ID ascending (code-unit order)
  const pPool = pool.filter(q => matchesFilters(q, topFilters, taxonomy));
  pPool.sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));

  if (pPool.length < count) {
    return {
      status: "infeasible",
      reasonCode: "insufficient_pool",
      message: `候选题目池数量不足（需要 ${count} 道，当前仅有 ${pPool.length} 道）`
    };
  }

  const totalSlotCount = slots.reduce((acc, s) => acc + s.count, 0);
  if (totalSlotCount > count) {
    return {
      status: "infeasible",
      reasonCode: "slot_conflict",
      message: `槽位需求题目总数 (${totalSlotCount}) 超过测试选卷总量 (${count})`
    };
  }

  // Step 2: Slot candidates
  const slotCandidates: Question[][] = [];
  for (let idx = 0; idx < slots.length; idx += 1) {
    const slot = slots[idx];
    let cands = pPool.filter(q => matchesFilters(q, slot.filters, taxonomy));
    if (seed !== undefined) {
      cands = shuffle(cands, `${setId}:slot:${slot.id}:${seed}`);
    }
    if (cands.length < slot.count) {
      return {
        status: "infeasible",
        reasonCode: "slot_insufficient_candidates",
        message: `槽位「${slot.id}」候选题目不足（需要 ${slot.count} 道，仅有 ${cands.length} 道）`
      };
    }
    slotCandidates.push(cands);
  }

  // Step 3: Backtracking search with constraint pruning and node budget
  const NODE_BUDGET = 5000;
  let nodeCount = 0;
  let exhausted = false;
  const assignedSlots: Question[][] = [];
  const usedIds = new Set<string>();

  function searchRemaining(
    needed: number,
    candidates: readonly Question[],
    currentSelection: readonly Question[],
    checkConstraints: boolean
  ): Question[] | null {
    if (needed === 0) {
      if (!checkConstraints || satisfiesConstraints(currentSelection, constraints)) {
        return [...currentSelection];
      }
      return null;
    }

    for (const chosen of combinations(candidates, needed)) {
      nodeCount += 1;
      if (nodeCount > NODE_BUDGET) {
        exhausted = true;
        return null;
      }
      const trial = [...currentSelection, ...chosen];
      if (!checkConstraints || satisfiesConstraints(trial, constraints)) {
        return trial;
      }
    }
    return null;
  }

  function searchSlots(slotIdx: number, checkConstraints: boolean): Question[] | null {
    nodeCount += 1;
    if (nodeCount > NODE_BUDGET) {
      exhausted = true;
      return null;
    }

    const flattened = assignedSlots.flat();
    const remainingTotal = count - flattened.length;
    if (checkConstraints && !canSatisfyConstraints(flattened, remainingTotal, constraints)) {
      return null;
    }

    if (slotIdx === slots.length) {
      const remainingNeeded = count - usedIds.size;
      let remainingCands = pPool.filter(q => !usedIds.has(q.id));
      if (remainingCands.length < remainingNeeded) return null;
      if (seed !== undefined) {
        remainingCands = shuffle(remainingCands, `${setId}:pool:${seed}`);
      }
      return searchRemaining(remainingNeeded, remainingCands, flattened, checkConstraints);
    }

    const slot = slots[slotIdx];
    const slotNeed = slot.count;
    const cands = slotCandidates[slotIdx].filter(q => !usedIds.has(q.id));
    if (cands.length < slotNeed) return null;

    for (const chosen of combinations(cands, slotNeed)) {
      const chosenIds = chosen.map(q => q.id);
      chosenIds.forEach(id => usedIds.add(id));
      assignedSlots.push(chosen);

      const result = searchSlots(slotIdx + 1, checkConstraints);
      if (result !== null) return result;
      if (exhausted) return null;

      assignedSlots.pop();
      chosenIds.forEach(id => usedIds.delete(id));
    }

    return null;
  }

  const solution = searchSlots(0, true);
  if (solution) {
    if (seed !== undefined) {
      const finalQuestions = shuffle(solution, `${setId}:order:${seed}`);
      return {
        status: "ok",
        questions: finalQuestions.map(q => withShuffledChoices(q, setId, seed))
      };
    }
    return { status: "ok", questions: solution };
  }

  if (exhausted) {
    return {
      status: "exhausted",
      message: "选题求解超出搜索节点预算，计算资源耗尽"
    };
  }

  // Check if failure was constraint-specific
  nodeCount = 0;
  assignedSlots.length = 0;
  usedIds.clear();
  const unconstrained = constraints.length > 0 ? searchSlots(0, false) : null;
  if (exhausted) {
    return {
      status: "exhausted",
      message: "选题求解超出搜索节点预算，计算资源耗尽"
    };
  }
  if (unconstrained !== null) {
    return {
      status: "infeasible",
      reasonCode: "constraint_violation",
      message: "题目组合无法满足题型、难度或风格等分布约束"
    };
  }

  return {
    status: "infeasible",
    reasonCode: "slot_conflict",
    message: "不同槽位之间的候选题目竞争导致无法同时满足"
  };
}

export function solveQuerySelection(
  pool: readonly Question[],
  query: SetSelectionQuery,
  taxonomy?: TaxonomyCatalog,
  seed?: string,
  setId = ""
): Question[] | null {
  const outcome = solveQuerySelectionDetailed(pool, query, taxonomy, seed, setId);
  return outcome.status === "ok" ? outcome.questions : null;
}

export function selectFixedSet(
  bundleQuestions: readonly Question[],
  selection: SetSelectionFixed,
  seed?: string,
  setId = ""
): Question[] {
  const byId = new Map(bundleQuestions.map(q => [q.id, q]));
  const questions: Question[] = [];
  for (const qid of selection.questions) {
    const q = byId.get(qid);
    if (!q) {
      throw new SelectionError(`Referenced question not found: ${qid}`);
    }
    questions.push(q);
  }

  const ordered =
    selection.order === "shuffle" && seed !== undefined ? shuffle(questions, `${setId}:order:${seed}`) : questions;

  return ordered.map(q => withShuffledChoices(q, setId, seed));
}

export function selectSetQuestions(bundle: SetBundle, seed: string, taxonomy?: TaxonomyCatalog): Question[] {
  if (!bundle.runnable) {
    throw new SelectionError(bundle.unavailableReason ?? "Set is not runnable");
  }

  if (bundle.set.selection.type === "fixed") {
    return selectFixedSet(bundle.questions, bundle.set.selection, seed, bundle.set.id);
  }

  const solution = solveQuerySelection(bundle.questions, bundle.set.selection, taxonomy, seed, bundle.set.id);
  if (!solution) {
    throw new SelectionError("No candidate questions satisfy selection criteria");
  }
  return solution;
}

export function selectRetry(
  availableQuestions: readonly Question[],
  ids: readonly string[],
  setId: string,
  seed = "retry"
): Question[] {
  const byId = new Map(availableQuestions.map(q => [q.id, q]));
  const questions: Question[] = [];
  for (const id of ids) {
    const q = byId.get(id);
    if (q) {
      questions.push(withShuffledChoices(q, setId, seed));
    }
  }
  return questions;
}

File: scripts/question_bank/selection.py
from __future__ import annotations

from dataclasses import dataclass
from itertools import combinations
import re
from typing import Any, Literal, TypeVar

from .models import TaxonomyRegistry

T = TypeVar("T")

SELECTION_ALGORITHM_VERSION = 1


@dataclass
class SelectionResult:
    status: Literal["ok", "infeasible", "exhausted"]
    questions: list[dict[str, Any]] | None = None
    reason_code: str | None = None
    message: str | None = None

    @property
    def is_ok(self) -> bool:
        return self.status == "ok"


def _int32(val: int) -> int:
    val = val & 0xFFFFFFFF
    if val >= 0x80000000:
        val -= 0x100000000
    return val


def _uint32(val: int) -> int:
    return val & 0xFFFFFFFF


def _imul(a: int, b: int) -> int:
    return _int32(_int32(a) * _int32(b))


def hash_seed(seed: str) -> int:
    h = 2166136261
    for byte in seed.encode("utf-8"):
        h = _uint32(h ^ byte)
        h = _uint32(_imul(h, 16777619))
    return h


def create_random(seed: str):
    state = hash_seed(seed)

    def next_float() -> float:
        nonlocal state
        state = _uint32(state + 0x6D2B79F5)
        value = _int32(state)
        v1 = _imul(value ^ (_uint32(value) >> 15), value | 1)
        v2 = _imul(v1 ^ (_uint32(v1) >> 7), v1 | 61)
        value = _int32(v1 ^ _int32(v1 + v2))
        return _uint32(value ^ (_uint32(value) >> 14)) / 4294967296.0

    return next_float


def shuffle(input_list: list[T], seed: str) -> list[T]:
    result = list(input_list)
    rnd = create_random(seed)
    for index in range(len(result) - 1, 0, -1):
        target = int(rnd() * (index + 1))
        result[index], result[target] = result[target], result[index]
    return result


def _matches_any_or_all(item_values: list[str], criterion: dict[str, list[str]]) -> bool:
    if "any" in criterion:
        target_set = set(criterion["any"])
        return any(v in target_set for v in item_values)
    if "all" in criterion:
        target_set = set(criterion["all"])
        return target_set.issubset(set(item_values))
    return True


def matches_filters(
    question: dict[str, Any],
    filters: dict[str, Any],
    taxonomy: TaxonomyRegistry | None = None,
) -> bool:
    if not filters:
        return True

    # 1. Topics
    if "topics" in filters:
        q_topics = question.get("topics")
        if not q_topics or not isinstance(q_topics, list):
            return False
        topic_crit = filters["topics"]
        if "any" in topic_crit:
            expanded: set[str] = set()
            for t in topic_crit["any"]:
                if taxonomy:
                    expanded.update(taxonomy.get_topic_descendants_and_self(t))
                else:
                    expanded.add(t)
            if not any(t in expanded for t in q_topics):
                return False
        elif "all" in topic_crit:
            for t in topic_crit["all"]:
                expanded = taxonomy.get_topic_descendants_and_self(t) if taxonomy else {t}
                if not any(qt in expanded for qt in q_topics):
                    return False

    # 2. Concepts
    if "concepts" in filters:
        q_concepts = question.get("concepts")
        if not q_concepts or not isinstance(q_concepts, list):
            return False
        if not _matches_any_or_all(q_concepts, filters["concepts"]):
            return False

    # 3. Objectives
    if "objectives" in filters:
        q_objs = question.get("objectives")
        if not q_objs or not isinstance(q_objs, list):
            return False
        if not _matches_any_or_all(q_objs, filters["objectives"]):
            return False

    # 4. Related Pages
    if "related_pages" in filters:
        q_pages = question.get("related_pages")
        if not q_pages or not isinstance(q_pages, list):
            return False
        if not _matches_any_or_all(q_pages, filters["related_pages"]):
            return False

    # 5. Types
    if "types" in filters:
        q_type = question.get("type")
        if q_type not in filters["types"]:
            return False

    # 6. Cognitive Levels
    if "cognitive_levels" in filters:
        q_cog = question.get("cognitive_level")
        if q_cog not in filters["cognitive_levels"]:
            return False

    # 7. Styles
    if "styles" in filters:
        q_style = question.get("style")
        if q_style not in filters["styles"]:
            return False

    # 8. Question IDs
    if "question_ids" in filters:
        q_id = question.get("id")
        if q_id not in filters["question_ids"]:
            return False

    # 9. Difficulty
    if "difficulty" in filters:
        q_diff = question.get("difficulty")
        if q_diff is None or not isinstance(q_diff, int):
            return False
        diff_crit = filters["difficulty"]
        min_diff = diff_crit.get("min")
        max_diff = diff_crit.get("max")
        if min_diff is not None and q_diff < min_diff:
            return False
        if max_diff is not None and q_diff > max_diff:
            return False

    return True


def satisfies_constraints(selected: list[dict[str, Any]], constraints: list[dict[str, Any]]) -> bool:
    for c in constraints:
        field = c.get("field")
        values = c.get("values", [])
        min_val = c.get("min")
        max_val = c.get("max")

        matching_count = sum(1 for q in selected if q.get(field) in values)
        if min_val is not None and matching_count < min_val:
            return False
        if max_val is not None and matching_count > max_val:
            return False
    return True


def can_satisfy_constraints(
    selected: list[dict[str, Any]],
    remaining_count: int,
    constraints: list[dict[str, Any]],
) -> bool:
    for c in constraints:
        field = c.get("field")
        values = c.get("values", [])
        min_val = c.get("min")
        max_val = c.get("max")
        matching = sum(1 for q in selected if q.get(field) in values)
        if max_val is not None and matching > max_val:
            return False
        if min_val is not None and matching + remaining_count < min_val:
            return False
    return True


def solve_query_selection_detailed(
    pool: list[dict[str, Any]],
    query: dict[str, Any],
    taxonomy: TaxonomyRegistry | None = None,
    seed: str | None = None,
    set_id: str = "",
) -> SelectionResult:
    count = query.get("count", 0)
    top_filters = query.get("filters", {})
    slots = query.get("slots", [])
    constraints = query.get("constraints", [])

    # Step 1: Filter pool by top-level filters and sort by id
    p_pool = [q for q in pool if matches_filters(q, top_filters, taxonomy)]
    p_pool.sort(key=lambda q: str(q.get("id", "")))

    if len(p_pool) < count:
        return SelectionResult(
            status="infeasible",
            reason_code="insufficient_pool",
            message=f"候选题目池数量不足（需要 {count} 道，当前仅有 {len(p_pool)} 道）",
        )

    # Step 2: Slot candidates
    slot_candidates: list[list[dict[str, Any]]] = []
    for s in slots:
        cands = [q for q in p_pool if matches_filters(q, s.get("filters", {}), taxonomy)]
        if seed is not None:
            cands = shuffle(cands, f"{set_id}:slot:{s.get('id')}:{seed}")
        slot_candidates.append(cands)

    total_slot_count = sum(s.get("count", 0) for s in slots)
    if total_slot_count > count:
        return SelectionResult(
            status="infeasible",
            reason_code="slot_conflict",
            message=f"槽位需求题目总数 ({total_slot_count}) 超过测试选卷总量 ({count})",
        )

    for idx, s in enumerate(slots):
        req = s.get("count", 0)
        available = len(slot_candidates[idx])
        if available < req:
            return SelectionResult(
                status="infeasible",
                reason_code="slot_insufficient_candidates",
                message=f"槽位「{s.get('id')}」候选题目不足（需要 {req} 道，仅有 {available} 道）",
            )

    # Step 3: Backtracking search with constraint pruning and node budget
    NODE_BUDGET = 5000
    node_count = 0
    exhausted = False
    assigned_slots: list[list[dict[str, Any]]] = []
    used_ids: set[str] = set()

    def search_slots(slot_idx: int, check_constraints: bool) -> list[dict[str, Any]] | None:
        nonlocal node_count, exhausted
        node_count += 1
        if node_count > NODE_BUDGET:
            exhausted = True
            return None

        flattened = [q for sublist in assigned_slots for q in sublist]
        remaining_total = count - len(flattened)
        if check_constraints and not can_satisfy_constraints(flattened, remaining_total, constraints):
            return None

        if slot_idx == len(slots):
            remaining_needed = count - len(used_ids)
            remaining_cands = [q for q in p_pool if str(q.get("id")) not in used_ids]
            if len(remaining_cands) < remaining_needed:
                return None
            if seed is not None:
                remaining_cands = shuffle(remaining_cands, f"{set_id}:pool:{seed}")
            return search_remaining(remaining_needed, remaining_cands, flattened, check_constraints)

        slot = slots[slot_idx]
        slot_need = slot.get("count", 0)
        cands = [q for q in slot_candidates[slot_idx] if str(q.get("id")) not in used_ids]
        if len(cands) < slot_need:
            return None

        for chosen in combinations(cands, slot_need):
            chosen_ids = {str(q.get("id")) for q in chosen}
            used_ids.update(chosen_ids)
            assigned_slots.append(list(chosen))

            result = search_slots(slot_idx + 1, check_constraints)
            if result is not None:
                return result
            if exhausted:
                return None

            assigned_slots.pop()
            used_ids.difference_update(chosen_ids)

        return None

    def search_remaining(
        needed: int,
        candidates: list[dict[str, Any]],
        current_selection: list[dict[str, Any]],
        check_constraints: bool,
    ) -> list[dict[str, Any]] | None:
        nonlocal node_count, exhausted
        if needed == 0:
            if not check_constraints or satisfies_constraints(current_selection, constraints):
                return list(current_selection)
            return None

        for chosen in combinations(candidates, needed):
            node_count += 1
            if node_count > NODE_BUDGET:
                exhausted = True
                return None
            trial = current_selection + list(chosen)
            if not check_constraints or satisfies_constraints(trial, constraints):
                return trial

        return None

    # First attempt: standard search with constraints
    solution = search_slots(0, check_constraints=True)
    if solution is not None:
        if seed is not None:
            final_questions = shuffle(solution, f"{set_id}:order:{seed}")
            result_questions: list[dict[str, Any]] = []
            for q in final_questions:
                q_copy = dict(q)
                if (
                    q_copy.get("choice_order") == "shuffle"
                    and "choices" in q_copy
                    and isinstance(q_copy["choices"], list)
                ):
                    q_copy["choices"] = shuffle(q_copy["choices"], f"{set_id}:{q_copy['id']}:choices:{seed}")
                result_questions.append(q_copy)
            return SelectionResult(status="ok", questions=result_questions)
        return SelectionResult(status="ok", questions=solution)

    if exhausted:
        return SelectionResult(
            status="exhausted",
            message="选题求解超出搜索节点预算，计算资源耗尽",
        )

    # Infeasible: check if failure was due to constraints or slot conflict
    node_count = 0
    assigned_slots.clear()
    used_ids.clear()
    unconstrained = search_slots(0, check_constraints=False) if constraints else None
    if exhausted:
        return SelectionResult(
            status="exhausted",
            message="选题求解超出搜索节点预算，计算资源耗尽",
        )
    if unconstrained is not None:
        return SelectionResult(
            status="infeasible",
            reason_code="constraint_violation",
            message="题目组合无法满足题型、难度或风格等分布约束",
        )

    return SelectionResult(
        status="infeasible",
        reason_code="slot_conflict",
        message="不同槽位之间的候选题目竞争导致无法同时满足",
    )


def solve_query_selection(
    pool: list[dict[str, Any]],
    query: dict[str, Any],
    taxonomy: TaxonomyRegistry | None = None,
    seed: str | None = None,
    set_id: str = "",
) -> list[dict[str, Any]] | None:
    res = solve_query_selection_detailed(pool, query, taxonomy, seed, set_id)
    return res.questions if res.status == "ok" else None

